ipam/pool_test.go

```diff
@@ -0,0 +1,1007 @@
+					poolId := "10.0.0.0/16"
+					as.Pools[poolId] = &addressPool{
+						Id:       poolId,
+						RefCount: 0,
+					}
+					ap, err := as.requestPool(poolId, "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal(poolId))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When pool is in use", func() {
+				It("Should raise an error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "10.0.0.0/16"
+					as.Pools[poolId] = &addressPool{
+						Id:       poolId,
+						RefCount: 1,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When pool is ipv4 and ipv6 is wanted", func() {
+				It("Should raise an error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "10.0.0.0/16"
+					as.Pools[poolId] = &addressPool{
+						Id:     poolId,
+						IsIPv6: false,
+					}
+					ap, err := as.requestPool("", "", nil, true)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When IfName is not match", func() {
+				It("Should raise an error", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "10.0.0.0/16"
+					ifName := "eth0"
+					as.Pools[poolId] = &addressPool{
+						Id:     poolId,
+						IfName: ifName,
+					}
+					options := map[string]string{}
+					options[OptInterfaceName] = "en0"
+					ap, err := as.requestPool("", "", options, false)
+					Expect(err).To(Equal(errNoAvailableAddressPools))
+					Expect(ap).To(BeNil())
+				})
+			})
+
+			Context("When addressSpace has one pool available", func() {
+				It("Should return the pool", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					poolId := "10.0.0.0/16"
+					as.Pools[poolId] = &addressPool{
+						Id: poolId,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal(poolId))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When addressSpace has pools with different priority", func() {
+				It("Should return the pool with the highest priority", func() {
+					as := &addressSpace{
+						Pools: map[string]*addressPool{},
+					}
+					as.Pools["10.0.0.0/16"] = &addressPool{
+						Id:       "10.0.0.0/16",
+						Priority: 1,
+					}
+					as.Pools["10.1.0.0/16"] = &addressPool{
+						Id:       "10.1.0.0/16",
+						Priority: 2,
+					}
+					ap, err := as.requestPool("", "", nil, false)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(ap.Id).To(Equal("10.1.0.0/16"))
+					Expect(ap.RefCount).To(Equal(1))
+				})
+			})
+
+			Context("When addressSpace has pools with different addresses", func() {
+				It("Should return the pool with the highest number of addresses", func() {
```

Comment:
can return any pool..what do you mean by highest number of addresses? Even in your testcase, all pool has equal numnber of addresses /16

Reply:
The `Addresses` of pool 10.0.0.0/16 is empty, but the `Addresses` of pool 10.1.0.0/16 contains `10.1.0.1/16`, so that pool 10.1.0.0/16 should be return.
`as.Pools["10.1.0.0/16"].Addresses["10.1.0.1/16"] = &addressRecord{}`

Reply:
ah got it